File: 1833-find-the-highest-altitude/find-the-highest-altitude.py
class Solution:
    def largestAltitude(self, gain: list[int]) -> int:
        # WHAT: Set the initial maximum altitude and running current altitude tracking variables to 0.
        # WHY: The trip begins at an altitude reference point of 0, which serves as our baseline.
        max_altitude = 0
        current_altitude = 0
        
        # WHAT: Walk through each point's change value within the gain list.
        for g in gain:
            # Update our running altitude calculation with the step's climb or descent value
            current_altitude += g
            
            # WHAT: Compare the current altitude to our peak recorded value, saving the higher result.
            # WHY: This records the highest overall peak achieved at any stage of the journey.
            if current_altitude > max_altitude:
                max_altitude = current_altitude
                
        return max_altitude

File: 1833-find-the-highest-altitude/find-the-highest-altitude.java
class Solution {
    public int largestAltitude(int[] gain) {
        // WHAT: Initialize the maximum altitude reached and the running current altitude to 0.
        // WHY: The biker starts at point 0 with an altitude of 0. We track both the peak height and our current position.
        int maxAltitude = 0;
        int currentAltitude = 0;
        
        // WHAT: Loop through each net change value in the gain array sequentially.
        // WHY: Each element represents the relative vertical shift between consecutive points.
        for (int g : gain) {
            // Update the running total altitude by adding the current net change
            currentAltitude += g;
            
            // WHAT: Update maxAltitude if the newly calculated altitude is higher than the previous peak.
            // WHY: This ensures we capture the absolute highest coordinate reached at any point during the trip.
            maxAltitude = Math.max(maxAltitude, currentAltitude);
        }
        
        // WHAT: Return the highest tracked altitude point.
        return maxAltitude;
    }
}